Realm/Realm/Sync/ErrorHandling/AutomaticRecoveryHandler.cs

```diff
@@ -0,0 +1,96 @@
+    /// In this situation the user can choose between two <see cref="Fallback"/> strategies: <see cref="Fallback.Manual"/> and <see cref="Fallback.DiscardLocal"/>.
+    /// The latter automatically discards all the local changes and uses the latest realm that is available on the remote sync server.
+    /// While the former simply calls <see cref="ManualResetFallback"/> to let the user manually handle the client reset.
+    /// You can read more about the merge rules at <see href="https://docs.mongodb.com/realm/sdk/dotnet/advanced-guides/client-reset/">Client Resets - .NET SDK</see>.
+    /// The automatic recovery mechanism will create write transactions meaning that all the changes that take place
+    /// are properly propagated through the standard Realm's change notifications.
+    /// The <see cref="AutomaticRecoveryHandler"/> strategy supplies three callbacks: <see cref="OnBeforeReset"/>, <see cref="OnAfterReset"/> and <see cref="ManualResetFallback"/>.
+    /// The first two are invoked just before and after the client reset has happened,
+    /// while the last one will be invoked in case an error occurs during the automated process and the system needs to fallback to a manual mode.
+    /// The overall recommendation for using this strategy is that using the three available callbacks should only be considered when:
+    /// 1. the user needs to be notified (in <see cref="OnBeforeReset"/>) of an incoming potential data loss of unsynced data as a result of a merge
+    ///    or a complete discard of local changes
+    /// 2. the user needs to be notified (in <see cref="OnAfterReset"/>) that the reset process has completed
+    /// 3. advanced use cases for data-sensitive applications where the developer wants to recover in the most appropriate way the unsynced data
+    /// 4. backup the whole realm before the client reset happens (in <see cref="OnBeforeReset"/>). Such backup could, for example, be used to restore the unsynced data (see 3.)
+    /// </remarks>
+    /// <seealso href="https://docs.mongodb.com/realm/sdk/dotnet/advanced-guides/client-reset/">Client Resets - .NET SDK</seealso>
+    public sealed class AutomaticRecoveryHandler : ClientResetHandlerBase
+    {
+        /// <summary>
+        /// Fallback strategies for the <see cref="AutomaticRecoveryHandler"/>.
+        /// </summary>
+        public enum Fallback
+        {
+            /// <summary>
+            /// If the automatic merge fails, manually handle the matter.
+            /// </summary>
+            Manual = 0,
+
+            /// <summary>
+            /// If the automatic merge fails, automatically discards all local changes and uses the latest realm that is available on the remote.
+            /// </summary>
+            DiscardLocal = 1,
+        }
+
+        internal Fallback FallbackStrategy { get; private set; }
```

Comment:
We can probably expose this publicly. Not super useful, but doesn't hurt either.